### ISSUE TITLE
gdb cannot access libcc1.so plugin

### ISSUE THREAD

**Issue by PowerUser64** [NixOS/nixpkgs#342904]
## Description
When using the `compile code` feature, GDB produces the following error:
```
Could not load libcc1.so: libcc1.so: cannot open shared object file: No such file or directory
```

### Steps To Reproduce
1. Make a simple program:
```c
#include <stdio.h>

int main() {
   printf("Hello, world!\n");
   return 0;
}
```
2. Compile it `gcc -O0 -g main.c`
3. Open the compiled result in GDB `gdb a.out`
4. Place a breakpoint on main `b main`
5. Run the program `r`
6. Try compiling and running some code from within GDB `compile code printf("Hello from GDB!\n")`
7. Profit

### Expected behavior
GDB should compile and execute the code written in the `compile code` command. You can test this on your system through [distrobox](https://github.com/89luca89/distrobox) if you want.

### Screenshots
![image](https://github.com/user-attachments/assets/3be610a7-2bc1-431b-b5c4-de5297decc77)

# Additional context

First, some background on `libcc1`.

- `libcc1` is a plugin for gdb that lets it compile C and C++ code through its `compile code` command.
- `libcc1` is part of gcc, and gdb is supposed to `dlopen()` it when `compile code` is called.
- gcc in nixpkgs has an option called `disableGdbPlugin` that causes `libcc1` to not be compiled, although I think it's `false` by default.

Now that we're on the same page, let's continue.

gdb [`dlopen()`s `libcc1.so`](https://github.com/bminor/binutils-gdb/blob/master/gdb/compile/compile-c-support.c#L111) ([defined here for C](https://github.com/bminor/binutils-gdb/blob/master/include/gcc-c-interface.h#L191-L194) and [here for c++](https://github.com/bminor/binutils-gdb/blob/master/include/gcc-cp-interface.h#L473-L476)), which it can't find under NixOS. I tried patching gdb with overrideAttrs (patch is below) to make it open `libcc1.so` from the nix store, including patching gcc to tell it to include the `lib` output in the `outputsToInstall` , which is the output that contains `libcc1.so`. This seems to get closer to working result, but produces the error `Could not find a compiler matching "^(x86_64|i.86)(-[^-]*)?-linux(-gnu[^-]*)?-gcc$"`, which I found is produced by `libcc1` by searching binary files in my nix store with ripgrep for `Could not find a compiler matching`.

Interestingly, I have noticed that there seem to be a few other places that I find `libcc1.so` in my nix store:
```
/nix/store/(sha256)-x86_64-unknown-linux-musl-gcc-13.3.0/lib/gcc/x86_64-unknown-linux-musl/13.3.0/plugin/libcc1plugin.so
/nix/store/(sha256)-gcc-13.2.0/lib/gcc/x86_64-unknown-linux-gnu/13.2.0/plugin/libcc1plugin.so
```
These are in addition to the path that I'm using in my patch below at `/nix/store/(sha256)-gcc-13.2.0-lib/lib/libcc1.so` (aka `${pkgs.gcc-unwrapped.lib}/lib/libcc1.so`). You'll notice that these places call the file `libcc1plugin.so` rather than `libcc1.so`, so I'm not convinced it is the same file or a valid replacement, but I wasn't able to test this because something in the build process replaces the `-linux-` part of the path with just `-l-` whenever I set gdb's  `GCC_C_FE_LIBCC` and `GCC_CP_FE_LIBCC` variables to point to that location. I'm not sure how to disable this behavior or what is causing it, so I don't know how to test this. (Although I presume it's nix.)

Are there any ideas as to what we can do to help it locate it? I would really like to get this functionality working. As I mentioned earlier, it works on other distros, including Arch Linux through distrobox.

```nix
            (gdb.overrideAttrs (new: {
              buildInputs = pkgs.gdb.buildInputs ++ [ gcc-unwrapped ];
              # nativeBuildInputs = pkgs.gdb.nativeBuildInputs ++ [ gcc libcxx ];
              env.NIX_LDFLAGS = lib.optionalString stdenv.hostPlatform.isLinux "-rpath ${lib.makeLibraryPath [ gcc-unwrapped ]}";
              dontPatchELF = true; # needed or nix will try to optimize the binary by removing "useless" rpath
              patches = [
                (pkgs.substituteAll {
                  gcc = "${((pkgs.gcc-unwrapped.override {
                      disableGdbPlugin = false;
                    }).overrideAttrs (oldAttrs: {
                    meta.outputsToInstall = oldAttrs.meta.outputsToInstall ++ [ "lib" ];
                  })).lib}";
                  # gcc = pkgs.gcc;
                  src =
                    (writeText "libcc-fix.patch" ''
                      diff --git a/include/gcc-c-interface.h b/include/gcc-c-interface.h
                      index 700d7483a4a..b5c800d3a44 100644
                      --- a/include/gcc-c-interface.h
                      +++ b/include/gcc-c-interface.h
                      @@ -191,7 +191,7 @@ struct gcc_c_context
                       /* The name of the .so that the compiler builds.  We dlopen this
                          later.  */

                      -#define GCC_C_FE_LIBCC libcc1.so
                      +#define GCC_C_FE_LIBCC @gcc@/lib/libcc1.so


                       /* The compiler exports a single initialization function.  This macro
                          holds its name as a symbol.  */
                      diff --git a/include/gcc-cp-interface.h b/include/gcc-cp-interface.h
                      index 15b911cb216..cf69c4f3d4f 100644
                      --- a/include/gcc-cp-interface.h
                      +++ b/include/gcc-cp-interface.h
                      @@ -473,7 +473,7 @@ struct gcc_cp_context
                       /* The name of the .so that the compiler builds.  We dlopen this
                          later.  */

                      -#define GCC_CP_FE_LIBCC libcc1.so
                      +#define GCC_CP_FE_LIBCC @gcc@/lib/libcc1.so

                       /* The compiler exports a single initialization function.  This macro
                          holds its name as a symbol.  */
                    '');
                })
              ];
            }))
```

### Notify maintainers

Pinging gcc and gdb maintainers since I'm not sure whether this should be fixed in GCC or GDB (could be both).

GDB:
- @nbp 
- @globin
- @lsix 

GCC:
- @Synthetica9 
- @vcunat
- @Ericson2314

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.10.9, NixOS, 24.11 (Vicuna), 24.11.20240910.1355a0c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(blake): `""`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/qpg5mwsind2hy35b9vpk6mx4jimnypw0-source`

```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by trofi** [comment]
Yeah, `nixpkgs` disabled `libcc1` with `pkgs/development/compilers/gcc/common/configure-flags.nix:    ++ lib.optional disableGdbPlugin "--disable-libcc1"`.

I think the primary motivation was only unexpected `libstdc++` references to the bootstrap compiler (as opposed to booted compiler, another negative side-effect of `--disable-bootstrap`): https://github.com/NixOS/nixpkgs/pull/216237/commits/24b07fc9e5216d8ae22205c2f8c688e2738d8502

**Comment by trofi** [comment]
I think `disableGdbPlugin = true;` specifically comes directly from `pkgs/stdenv/linux/default.nix`:

```nix
  commonGccOverrides = {
    # Use a deterministically built compiler
    # see https://github.com/NixOS/nixpkgs/issues/108475 for context
    reproducibleBuild = true;
    profiledCompiler = false;

    # It appears that libcc1 (which is not a g++ plugin; it is a gdb plugin) gets linked against
    # the libstdc++ from the compiler that *built* g++, not the libstdc++ which was just built.
    # This causes a reference chain from stdenv to the bootstrapFiles:
    #
    #   stdenv -> gcc-lib -> xgcc-lib -> bootstrapFiles
    #
    disableGdbPlugin = true;
  };
```

We apply this override to the multiple stages (probably all that build `gcc`, not just the early ones).
